drivers/gpu/drm/rp1/rp1-dsi/rp1_dsi_dsi.c

```diff
@@ -1124,19 +1124,23 @@ static uint8_t dphy_get_div(u32 refclk_khz, u32 vco_freq_khz, u32 *ptr_m, u32 *p
 	 * In practice, given a 50MHz reference clock, it can produce any
 	 * multiple of 10MHz, 11.1111MHz, 12.5MHz, 14.286MHz or 16.667MHz
 	 * with < 1% error for all frequencies above 495MHz.
+	 *
+	 * vco_freq should be set to the lane bit rate (not the MIPI clock
+	 * which is half of this). These frequencies are now measured in Hz.
+	 * They should fit within u32, but u64 is needed for calculations.
 	 */
 
-	static const u32 REF_DIVN_MAX = 40000u;
-	static const u32 REF_DIVN_MIN =  5000u;
-	u32 best_n, best_m, best_err = 0x7fffffff;
-	unsigned int n;
+	static const u64 REF_DIVN_MAX = 40000000;
+	static const u64 REF_DIVN_MIN =  5000000;
+	unsigned int n, best_n, best_m;
+	u64 best_err = vco_freq;
 
-	for (n = 1 + refclk_khz / REF_DIVN_MAX; n * REF_DIVN_MIN <= refclk_khz && n < 100; ++n) {
-		u32 half_m = (n * vco_freq_khz + refclk_khz) / (2 * refclk_khz);
+	for (n = 1 + refclk / REF_DIVN_MAX; n * REF_DIVN_MIN <= refclk && n < 100; ++n) {
```

Comment:
Need to be careful to ensure this can't be built on a 32bit kernel as it'll error out on 64bit division. I'd hope a Kconfig change can protect against that.

Reply:
Can a 32-bit kernel not do u64 division? I can probably re-work the arithmetic to avoid it but it would look messier.

Reply:
It can, but not as a raw divide.

`do_div`, or you've got `DIV_ROUND_CLOSEST`, `DIV_ROUND_DOWN`, and `DIV_ROUND_UP` type macros (see include/linux/math.h)

Reply:
I'm sure it can - the compiler will generate calls to helper functions - provided at least one of the types is 64-bit.

Reply:
The kernel build checks have all been successful.

Reply:
`div_u64()` looks like the one to use.

Reply:
CONFIG_DRM_RP1_DSI is only set in arch/arm64/configs/bcm2711_defconfig and arch/arm64/configs/bcm2712_defconfig, so no 32bit build is trying to compile it.

If it were to try, then it typically fails with errors such as 
```
ERROR: "__aeabi_ldivmod" [/home/pi/media_build/v4l/mn88436.ko] undefined!
ERROR: "__aeabi_uldivmod" [/home/pi/media_build/v4l/mn88436.ko] undefined!
```
Certainly for upstreaming you'll get the random config builds managing to try building it on 32bit and failing. Kconfig should block that.

My 32bit build + MFD_RP1 and DRM_RP1_DSI has just failed with 
```
ERROR: modpost: "msi_domain_set_affinity" [drivers/mfd/rp1.ko] undefined!
ERROR: modpost: "irq_domain_deactivate_irq" [drivers/mfd/rp1.ko] undefined!
ERROR: modpost: "irq_domain_activate_irq" [drivers/mfd/rp1.ko] undefined!
```
instead, so bigger fish to fry.

More an observation than requiring fixing immediately.

Reply:
(Largely to myself)
Apparently this is because the kernel doesn't use libgcc and doesn't include its own versions of these functions. The functions declared in math64.h should be used instead.